### DISCUSSION TITLE
🚀 Feature: Realtime event triggers

### DISCUSSION THREAD

**Discussion by Meldiron** [appwrite/appwrite#11505] [Ideas]
### 🔖 Feature description

We should add new event types that can trigger a function. Alongside existing ones, we should add new ones for real-time events, such as user connection or disconnection.

### 🎤 Pitch

Currently, I am trying to build a real-time game and I need to know when a user disconnect, so I can do cleanup on the client-side of other players - remove a character of a player who disconnected. I don't think this is possible using current real-time features.

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/HEAD/CODE_OF_CONDUCT.md)

**Comment by eldadfux** [comment]
@TorstenDittmann do you think it will make sense to add a callback for when the user disconnect?

**Comment by TorstenDittmann** [comment]
Not sure how the channel and permissions should look like for this 🤔

**Comment by eldadfux** [comment]
> Not sure how the channel and permissions should look like for this 🤔

Maybe for start, just allow callbacks in the SDK level, no implementation on the server side, or is that already supported?

**Comment by weepy** [comment]
I also need some user presence detection. Is there any movement on this ?

**Comment by punitdiwan** [comment]
@TorstenDittmann  Any update? I am planning to set up a chat server and need the user's Online / Offline presence support. Is there any workaround you can suggest to how to handle the user's disconnection or point to some resource?    

I think this is really helpful for real-time chat apps. Rest can be handled by real-time updates from collections. But the socket connects and disconnect event is missing which can be added.   

Waiting for an update.

Regards,
Punit Diwan

**Comment by pepperoni21** [comment]
Any update on this? The mentionned issue is not about presence system, it's about client-side events

**Comment by kecsot** [comment]
Yes, **it would be nice to handle clientside connects/disconnects**. I think it's an important action when we use Realtime data. There is any solution to handle that? 

I think the possible solutions:

1. Allow to use custom realtime implementation. (Currently it's `private`) ( https://github.com/appwrite/sdk-for-web/blob/master/src/client.ts#L182 ) It can gives a lot of options to change default realtime implementation. Like change `getTimeout` logic, write attempt limiter, etc.
2. Create Callback  like:`onRealtimeConnectionChanged()`
3. Use channels ( https://github.com/appwrite/sdk-for-web/issues/40#issuecomment-1259837819 )

**Comment by KimCalvin** [comment]
Is this feature being considered? I've also seen this feature request on the appwrite discord regularly.

**Comment by xuelink** [comment]
It is very old requested feature, but also it has still high-demand like [push-notifications feature](https://github.com/appwrite/appwrite/discussions/1623). 

I came here from firebase, when i was in DB, i solve presence issue by writing Cloud Function there. Anyways, my migration from firebase to appwrite almost completed except this issue. 

Is it possible way to solve it by writing an appwrite cloud function ? If yes, please link, mention, any referance, sample function or readings that i need. Because it seems like it will take time for this feature to be released. Big Thanks!

**Comment by xuelink** [comment]
[discord related link](https://discord.com/channels/564160730845151244/564160731327758347/1163772249363447818)
<img width="1416" alt="Screenshot 2023-10-17 at 19 01 15" src="https://github.com/appwrite/appwrite/assets/3280475/1798aee3-2902-4cce-8a56-b8cf2c9ed260">

Up!

**Comment by emilepavie** [comment]
any update on this ?? @stnguyen90

**Comment by xuelink** [comment]
I am also waiting an update about it as well.
